### PR TITLE
Customer account invoicing

### PR DESCRIPTION
**Pull request by gbova**
@nickpell  This PR is pretty big now. I added:

`CustomerInvoice` now includes the customer account and the list of subscriptions that are relevant to the invoice. Including those subscriptions ended up being necessary to apply credits, an they may be good to have if we want a way to remind billing admins of their plans/domains in the future.

`CustomerBillingRecord` to handle emailing the invoice and applying credits.

`CustomerInvoiceSummaryView` and `InvoicePdfView` as URLs to view a customer invoice and its PDF, respectively.

`CustomerInvoiceInterface` is to view/search for customer invoices, similar to Invoice and WireInvoice interfaces.

The invoicing process for customer-level accounts creates a CustomerInvoice and a CustomerBillingRecord now, so only one invoice will be created. There is a Customer Invoices tab under "Accounting Admin" to view/search for all customer invoices.

(I still need to write more tests for this.)

### CODE REVIEW

File: corehq/apps/accounting/migrations/0032_customerinvoice.py
Comment (nickpell) [inline corehq/apps/accounting/migrations/0032_customerinvoice.py]: It's good practice to squash multiple migrations in the same PR into one migration.  You can do this manually or use django's tools to help:
https://docs.djangoproject.com/en/1.11/topics/migrations/#migration-squashing
https://docs.djangoproject.com/en/1.11/ref/django-admin/#django-admin-squashmigrations

File: corehq/apps/accounting/models.py
Comment (nickpell) [inline corehq/apps/accounting/models.py]: Curious what you think about adding a property instead of a new database field?  Something along the lines of

```
@property
@memoized
def subscriptions(self):
    return self.account.subscription_set()
```

Comment (gbova) [reply corehq/apps/accounting/models.py]: `CustomerInvoice.subscriptions` is set when we decide what subscriptions to invoice in `invoicing.py` (based on the dates of the invoicing period, `should_create_invoice`, etc), and are specific to the invoice itself. I believe `account.subscription_set()` is a list of all subscriptions that have been associated with the account, including subscriptions that we would not want to invoice

File: corehq/apps/accounting/models.py
Comment (gbova) [reply corehq/apps/accounting/models.py]: @nickpell is there something else I should be using?

File: corehq/apps/accounting/views.py
Comment (gbova) [reply corehq/apps/accounting/views.py]: GitHub won't show me the line of code you're referencing, but I think you're talking about the `CustomerInvoiceSummaryView` class, which is the page you land on when you view an invoice (not its PDF)
screen shot 2018-07-06 at 1 35 31 pm

File: corehq/apps/accounting/migrations/0032_customerinvoice.py
Comment (gbova) [reply corehq/apps/accounting/migrations/0032_customerinvoice.py]: At what point can I delete the old migration files? Once this squashed migrations file has been deployed to everything?

File: corehq/apps/accounting/models.py
Comment (nickpell) [reply corehq/apps/accounting/models.py]: It should be safe to ignore this warning.  There isn't a library to get the python3 behavior.

For a full description of the differences of ```round```: http://python3porting.com/differences.html#rounding-behavior

File: corehq/apps/accounting/views.py
Comment (nickpell) [reply corehq/apps/accounting/views.py]: Sorry, I meant ```InvoicePdfView```.  I saw you defined this class and included it in the urls.py, but didn't see the url being used.

File: corehq/apps/accounting/migrations/0032_customerinvoice.py
Comment (nickpell) [reply corehq/apps/accounting/migrations/0032_customerinvoice.py]: Another point: when you do delete the squashed migrations, you should also remove the ```replaces``` field from the new migration at the same time.

File: corehq/apps/accounting/views.py
Comment (gbova) [reply corehq/apps/accounting/views.py]: Ah, that makes so much more sense!

It's used to view PDFs generated from customer-level invoices, specifically (after going through `CustomerInvoiceInterface` and `CustomerInvoiceSummaryView` you're going to get to `InvoicePdfView `).

The `customer_invoices` slug is in `invoice.html`:
```
{% if billing_record.invoice.is_customer_invoice %}
                                <td>
                                    <a href="{% url 'invoice_pdf_view' billing_record.pdf_data_id %}">
                                        View PDF
                                    </a>
                                </td>
                            {% else %}
                                <td>
                                    <a href="{% url 'domain_billing_statement_download' billing_record.invoice.get_domain billing_record.pdf_data_id %}">
                                        View PDF
                                    </a>
                                </td>
                            {% endif %}
```

File: corehq/apps/accounting/migrations/0032_customerinvoice.py
Comment (gbova) [reply corehq/apps/accounting/migrations/0032_customerinvoice.py]: So, now that I have that squashed migrations file, I should run migrations -> delete `replaces` -> run migrations again?

I haven't put this on staging, so I'm just going to do all this locally.

File: corehq/apps/accounting/models.py
Comment (nickpell) [reply corehq/apps/accounting/models.py]: > I believe account.subscription_set() is a list of all subscriptions that have been associated with the account, including subscriptions that we would not want to invoice

That's true, although I suppose you could add the relevant filters in this property.

How will this field be used?

File: corehq/apps/accounting/migrations/0032_customerinvoice.py
Comment (nickpell) [reply corehq/apps/accounting/migrations/0032_customerinvoice.py]: 👍

File: corehq/apps/accounting/views.py
Comment (nickpell) [reply corehq/apps/accounting/views.py]: 👍

File: corehq/apps/accounting/models.py
Comment (nickpell) [reply corehq/apps/accounting/models.py]: 👍

File: corehq/apps/accounting/views.py
Comment (nickpell) [reply corehq/apps/accounting/views.py]: small nit: it might be helpful for future devs to rename ```InvoicePdfView``` to ```CustomerInvoicePdfView``` or similar to clarify that this is not intended for subscription level invoices (not blocking merging)

**Review by nickpell — APPROVED** [review]
This is good to go from me, once the merge conflict with master is resolved.  Nice work!

**Review by nickpell — APPROVED** [review]

**Pull request merged**
